### PR TITLE
[Reland] Added reference tests to ReductionOpInfo

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64273

Reintroduced sample_inputs_prod and constrained the range of values for large reference tests.

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

Differential Revision: [D30672097](https://our.internmc.facebook.com/intern/diff/D30672097)

### CODE REVIEW

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp [hunk 1]
```diff
@@ -163,24 +163,29 @@ static void std_var_kernel_impl(TensorIterator& iter, int64_t correction, bool t
...
-      // NOLINTNEXTLINE(bugprone-argument-comment)
-      /*identity=*/1);
-  } else {
-    AT_DISPATCH_ALL_TYPES_AND_COMPLEX(iter.dtype(), "prod_cpu", [&] {
-      binary_kernel_reduce_vec(
         iter,
-        [=](scalar_t a, scalar_t b) -> scalar_t { return a * b; },
-        [=](Vectorized <scalar_t> a, Vectorized <scalar_t> b) { return a * b; },
+        [=](scalar_t a, scalar_t b)
+            __ubsan_ignore_undefined__ -> scalar_t { return a && b; },
```
Comment (albanD) [inline aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:173]: What does these guards do?
Why were they not needed before?

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -782,8 +784,8 @@ def _generate_reduction_inputs(device, dtype, requires_grad):
     """Generates input tensors for testing reduction operators"""
     yield make_tensor([], device, dtype, requires_grad=requires_grad)
     yield make_tensor([2], device, dtype, requires_grad=requires_grad)
-    yield make_tensor([2, 3], device, dtype, requires_grad=requires_grad, noncontiguous=True)
-    yield make_tensor([3, 2, 1, 5], device, dtype, requires_grad=requires_grad)
+    yield make_tensor([3, 5], device, dtype, requires_grad=requires_grad, noncontiguous=True)
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:787]: Why the change of size?

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -9288,6 +9336,37 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+            # FIXME: fix precision
+            DecorateInfo(toleranceOverride({
+                torch.float16: tol(atol=1e-05, rtol=1e-02),
+            }), 'TestReductions', 'test_noncontiguous_all'),
+            DecorateInfo(toleranceOverride({
+                torch.float16: tol(atol=1e-05, rtol=1e-02),
+            }), 'TestReductions', 'test_ref_small_input'),
+        ),
+        skips=(
+            # FIXME: prod does not support passing keepdim without passing dim
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:9358]: These comments should mention mean instead of prod?

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 1 above)
Comment (heitorschueroff) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:173]: They suppress undefined behavior error from https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html. These were needed because one of the builds was reporting undefined behavior on this kernel. This is used in other kernels as well.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:787]: To keep the test cases small and prevent overflow with prod.

Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:787]: You go from 2, 3 to 3, 5. That doesn't really make the test case smaller? :D

**Comment by mruberry** [comment]
fyi paralleltbb failure is unrelated and now resolved

**Comment by mruberry** [comment]
@albanD you're shepherding this, right?

**Comment by albanD** [comment]
@mruberry yes I can take a look.

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:787]: I made the [3, 2, 1, 5] -> [3, 2, 1, 2] smaller and [2, 3] -> [3, 5] bigger to compensate? Haha, these are arbitrary numbers at the end of the day.

**Comment by heitorschueroff** [comment]
@albanD This should be ready for another review

File: torch/testing/_internal/common_methods_invocations.py (hunk 3 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py]: Addressed

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -5720,6 +5723,53 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
...
+        keys = set(kwargs.keys())
+
+        dim = kwargs.pop('dim', None)
+        keepdim = kwargs.pop('keepdim', False)
+
+        if 'dim' in keys:
+            if x.ndim == 0:
+                # NumPy reductions don't accept dim=0 for scalar inputs
+                for i in dim if isinstance(dim, tuple) else (dim,):
+                    assert i in {0, -1}
```
[the remark this answers is not in the document]
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py]: Changed the logic so the error is delegated to the actual operation.

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:788]: Yes they are. That's why I was surprised you changed them.

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:173]: Is that because some of the new tests that are added?
And is it fine to ignore these?

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -188,6 +188,8 @@ def _np(t):
                 return tuple(map(to_numpy, x))
             elif isinstance(x, dict):
                 return {k: to_numpy(v) for k, v in x.items()}
+            elif isinstance(x, torch.dtype):
+                return torch.empty(0, dtype=x).numpy().dtype
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:192]: Interesting move :D
Maybe we can re-use: https://github.com/pytorch/pytorch/blob/421d8f86b6def536df18371a5da2f5df4de6e262/torch/testing/_internal/common_utils.py#L455 ?

**Review by albanD — COMMENTED** [review]
I'm not super happy with all the tolerance overrides but I think the PR is good :)

File: aten/src/ATen/native/cpu/ReduceOpsKernel.cpp (hunk 1 above)
Comment (heitorschueroff) [reply aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:173]: I assume it's because of the new tests since that's all that changed. I think it's safe to ignore it, we have other kernels doing the same. Ideally we should look into why the undefined behavior is happening and fix it though.

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py:192]: > I'm not super happy with all the tolerance overrides but I think the PR is good :)

Could you elaborate more on your concerns about the tolerance overrides? Do you think we should skip the tests instead? Or if it a concern with the numerical instability they expose?

Comment (albanD) [reply torch/testing/_internal/common_methods_invocations.py:192]: If so many functions need to override the default value, maybe the default is not good?
Or we want to run these in double to make sure they are correct (like we do for gradcheck)?

Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py]: The problem is mostly with sum and prod for float16. I changed the decorator to override the tolerance for float16 for all tests. Some tests are specifically testing numerical stability so it's expected we need to override the tolerance, this is also a way to document numerical instability.

**Comment by heitorschueroff** [comment]
@albanD What else is this PR missing to get accepted?

File: torch/testing/_internal/common_methods_invocations.py (hunk 5 above)
Comment (heitorschueroff) [reply torch/testing/_internal/common_methods_invocations.py]: I changed the large reference tests from float32 to float64

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

**Comment by heitorschueroff** [comment]
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

**Review by albanD — APPROVED** [review]
Thanks for the update!
